rfcs/20181016-optimizer-unification.md

```diff
@@ -0,0 +1,398 @@
+- Add a `get_weights` and `set_weights` method, to retrieve (or set) the optimizer’s state as a list of numpy arrays -- this is necessary for compatibility with the Keras API.
+- Add ability to set the values of optimizer hyperparameters (i.e. the arguments that can be passed to the constructor) at any point in the lifetime of the optimizer, without having to reinstantiate it. In particular this includes the ability to change the value of the learning rate.
+- Add support for gradient clipping by norm and by value.
+- Disable reusing a single optimizer instance across multiple graphs.
+- Move the optimizer classes to `tf.keras.optimizers`, with revise signatures (see details below).
+
+
+---
+
+## Detailed Design
+
+### I - Add a get_config method on every optimizer:
+
+```python
+optimizer.get_config()
+```
+
+This method is already present on the Model class and every Layer class.
+
+**Returns:**
+- A JSON-serializable dictionary (does not contain any non-serializable data such as tensors) containing the configuration of the optimizer, i.e. its constructor arguments. For instance, for Adadelta, this would look like `{'learning_rate': 0.1, 'rho': 0.95, 'epsilon': 1e-8, 'name': 'my_optimizer'}`
+
+
+### II - Add a from_config class method on every optimizer (only needs a single implementation on the base class):
+
+```python
+optimizer = Adadelta.from_config(config)
+```
+
+This method is already present on the Model class and every Layer class. This method is required for Keras compatibility.
+
+**Args:**
+- config: A dictionary, containing the same keys as what gets returned by `get_config`.
+
+**Returns:**
+- An optimizer instance with the desired configuration, effectively a clone of the original optimizer (minus its state, i.e. its weight values).
+
+
+### III - Add a get_weights method on every optimizer (only needs a single implementation on the base class):
+
+```python
+optimizer.get_weights()
+```
+
+This method is already present on the Model class and every Layer class.
+
+**Returns:**
+- A flat list of Numpy arrays, in deterministic order, where each array represents the value of an internal weight of the optimizer (such as the momentum of a model weight).
+
+
+### IV - Add a set_weights method on every optimizer (only needs a single implementation on the base class):
+
+```python
+optimizer.set_weights(weights)
+```
+
+This method is already present on the Model class and every Layer class. This method is required for Keras compatibility.
+
+**Args:**
+- weights: A flat list of Numpy arrays, in deterministic order, same as returned by get_weights. Note that since the optimizer creates its internal weights to match the set of weights it is trying to optimize, set_weights would only match get_weights when the set of weights being optimized is equivalent. E.g.:
```

Comment:
It seems this interface does not support the case of taking model A, taking part of it as model B, and importing optimizer weights from model A into model B.